File: java/src/main/java/org/solutions/leetcode/easy/BinaryWatch.java
package org.solutions.leetcode.easy;

import java.util.ArrayList;
import java.util.List;

/**
 * <a href="https://leetcode.com/problems/binary-watch">401. Binary Watch</a>
 */

public class BinaryWatch {
    public List<String> readBinaryWatch(int turnedOn) {
        if (turnedOn > 8) return List.of();

        List<String> res = new ArrayList<>();

        List<List<Integer>> hourGroups = new ArrayList<>();
        List<List<Integer>> minuteGroups = new ArrayList<>();
        for (int i = 0; i <= 4; i++) {
            hourGroups.add(new ArrayList<>());
        }
        for (int i = 0; i <= 6; i++) {
            minuteGroups.add(new ArrayList<>());
        }

        for (int i = 0; i <= 11; i++) {
            hourGroups.get(Integer.bitCount(i)).add(i);
        }
        for (int i = 0; i <= 59; i++) {
            minuteGroups.get(Integer.bitCount(i)).add(i);
        }

        for (int i = 0; i <= Math.min(turnedOn, 4); i++) {
            int turnOnWithoutHour = turnedOn - i;
            if (turnOnWithoutHour >= 0 && turnOnWithoutHour <= 6) {
                for (int hour : hourGroups.get(i)) {
                    for (int minute : minuteGroups.get(turnOnWithoutHour)) {
                        res.add(String.format("%d:%02d", hour, minute));
                    }
                }
            }
        }

        return res;
    }
}

File: golang/solution/leetcode/easy/binary_watch.go
package easy

import (
	"fmt"
	"math/bits"
)

// https://leetcode.com/problems/binary-watch

func readBinaryWatch(turnedOn int) []string {
	var res []string

	for i := 0; i <= 11; i++ {
		hourBitCount := bits.OnesCount(uint(i))
		for j := 0; j <= 59; j++ {
			minuteBitCount := bits.OnesCount(uint(j))

			if hourBitCount+minuteBitCount == turnedOn {
				s := fmt.Sprintf("%d:%02d", i, j)
				res = append(res, s)
			}
		}
	}

	return res
}

func ReadBinaryWatchWrapper(turnedOn int) []string {
	return readBinaryWatch(turnedOn)
}
